File: ts/test/spec.test.ts
import { test, describe } from 'node:test'
import Assert from 'node:assert'
import Fs from 'node:fs'
import Path from 'node:path'

// Internal module paths, not the package entry: the corpus covers helpers
// such as `get` that are not on the public API surface. point.test.ts
// already reaches into dist/ the same way.
import * as Basic from '../dist/util/basic'
import * as DiffUtil from '../dist/diff'


// The shared corpus, driven by both stacks. See test/spec/README.md.
// __dirname is ts/dist-test at run time, so the repo root is two up.
const SPEC_DIR = Path.join(__dirname, '..', '..', 'test', 'spec')


type Case = {
  file: string
  line: number
  id: string
  fn: string
  args: any[]
  expect: any
  error: string
}


// One adapter per corpus `fn`. Shapes are chosen so the result serializes
// identically in Go: `omap` yields an ordered pair list because that is
// the only way to assert key order in a language whose maps have none.
const FN: Record<string, (a: any[]) => any> = {
  camelify: (a) => (Basic as any).camelify(a[0]),
  snakify: (a) => (Basic as any).snakify(a[0]),
  kebabify: (a) => (Basic as any).kebabify(a[0]),
  partify: (a) => (Basic as any).partify(a[0]),
  lcf: (a) => (Basic as any).lcf(a[0]),
  ucf: (a) => (Basic as any).ucf(a[0]),
  escre: (a) => (Basic as any).escre(a[0]),
  indent: (a) => (Basic as any).indent(a[0], a[1]),
  isbinext: (a) => (Basic as any).isbinext(a[0]),
  isbincontent: (a) => (Basic as any).isbincontent(a[0]),
  get: (a) => (Basic as any).get(a[0], a[1]),
  getx: (a) => (Basic as any).getx(a[0], a[1]),
  deep: (a) => (Basic as any).deep(...a),
  omap: (a) => Object.entries((Basic as any).omap(a[0])),
  template: (a) => (Basic as any).template(a[0], a[1], a[2]),
  names: (a) => 2 === a.length
    ? (Basic as any).names(a[0], a[1])
    : (Basic as any).names(a[0], a[1], a[2]),
  lines: (a) => (DiffUtil as any).lines(a[0]),
  lcs: (a) => (DiffUtil as any).lcs(a[0], a[1]),
}


function loadCases(): Case[] {
  const out: Case[] = []

  const files = Fs.readdirSync(SPEC_DIR)
    .filter((n) => n.endsWith('.tsv'))
    .sort()

  if (0 === files.length) {
    throw new Error('no .tsv files found in ' + SPEC_DIR)
  }

  for (const file of files) {
    // Corpus files are committed LF and .gitattributes keeps them that way,
    // but normalise anyway: a clone made before that entry existed, or a
    // zip download, still lands CRLF, and a stray \r on the last cell is a
    // baffling failure to debug (`want "error", got "error\r"`). The Go
    // runner does the same.
    const text = Fs.readFileSync(Path.join(SPEC_DIR, file), 'utf8')
      .replace(/\r\n/g, '\n')
    const rows = text.split('\n')
    let header: string[] | null = null

    for (const [i, row] of rows.entries()) {
      if ('' === row.trim() || row.startsWith('#')) {
        continue
      }

      const cells = row.split('\t')

      if (null == header) {
        header = cells
        const want = ['id', 'fn', 'args', 'expect', 'error']
        Assert.deepStrictEqual(header, want,
          `${file}: header is ${JSON.stringify(header)}, want ${JSON.stringify(want)}`)
        continue
      }

      // A short row means a trailing empty cell was trimmed by an editor;
      // pad rather than crash on an out-of-range index.
      while (cells.length < 5) {
        cells.push('')
      }
      Assert.equal(cells.length, 5,
        `${file}:${i + 1}: ${cells.length} cells, want 5`)

      const error = cells[4]

      out.push({
        file,
        line: i + 1,
        id: cells[0],
        fn: cells[1],
        args: JSON.parse(cells[2]),
        expect: '' === error ? JSON.parse(cells[3]) : undefined,
        error,
      })
    }
  }

  return out
}


// Canonical JSON with object keys sorted, matching what Go's
// json.Marshal produces. Arrays keep their order — they have a
// meaningful one, and that is what the ordering cases rely on.
function canon(val: any): string {
  return JSON.stringify(sorted(val))
}

function sorted(val: any): any {
  if (null == val || 'object' !== typeof val) {
    // undefined is a miss, and both stacks report misses as null.
    return undefined === val ? null : val
  }
  if (Array.isArray(val)) {
    return val.map(sorted)
  }
  const out: any = {}
  for (const key of Object.keys(val).sort()) {
    out[key] = sorted(val[key])
  }
  return out
}


describe('spec-corpus', () => {

  const cases = loadCases()

  // A corpus that silently shrinks to nothing would pass. Assert it is
  // actually populated, and report the count so the two stacks can be
  // compared at a glance.
  test('corpus-loaded', () => {
    Assert.ok(100 < cases.length,
      `only ${cases.length} cases loaded from ${SPEC_DIR}`)
    console.log(`spec corpus: ${cases.length} cases`)
  })

  // Unknown `fn` is a failure, not a skip: a corpus entry one stack
  // ignores is exactly the divergence this suite exists to catch.
  test('all-fns-dispatched', () => {
    const missing = [...new Set(cases.map((c) => c.fn))]
      .filter((fn) => !(fn in FN))
      .sort()
    Assert.deepStrictEqual(missing, [],
      'corpus uses undispatched fns: ' + missing.join(', '))
  })

  test('ids-unique', () => {
    const seen = new Map<string, string>()
    for (const c of cases) {
      const key = c.file + ':' + c.id
      Assert.ok(!seen.has(key), `duplicate id ${key}`)
      seen.set(key, c.file)
    }
  })

  for (const c of cases) {
    test(`${c.file}/${c.id}`, () => {
      const where = `${c.file}:${c.line} ${c.id}`

      if ('' !== c.error) {
        Assert.throws(
          () => FN[c.fn](c.args),
          (err: any) => {
            Assert.ok(String(err.message).includes(c.error),
              `${where}: message ${JSON.stringify(err.message)} ` +
              `does not contain ${JSON.stringify(c.error)}`)
            return true
          },
          `${where}: expected a throw containing ${JSON.stringify(c.error)}`)
        return
      }

      const actual = FN[c.fn](c.args)
      Assert.equal(canon(actual), canon(c.expect), where)
    })
  }

})

File: go/spec_test.go
package jostraca

import (
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"reflect"
	"sort"
	"strings"
	"testing"
)

// The shared TSV corpus in test/spec, driven by both implementations.
// ts/test/spec.test.ts reads the same files and asserts the same
// expectations; see test/spec/README.md for the format.
//
// Tests run with the package directory as the working directory, so the
// corpus is one level up.
const specDir = "../test/spec"

type specCase struct {
	file   string
	line   int
	id     string
	fn     string
	args   []any
	expect any
	errSub string
}

// One adapter per corpus `fn`. These translate corpus JSON into native
// calls, which is where JSON's float64-everything meets Go's typed
// signatures -- `indent` takes an int count, `isbincontent` takes bytes.
var specFns = map[string]func(a []any) (any, error){
	"camelify": func(a []any) (any, error) { return Camelify(a[0]), nil },
	"snakify":  func(a []any) (any, error) { return Snakify(a[0]), nil },
	"kebabify": func(a []any) (any, error) { return Kebabify(a[0]), nil },
	"partify":  func(a []any) (any, error) { return Partify(a[0]), nil },
	"lcf":      func(a []any) (any, error) { return LCF(a[0]), nil },
	"ucf":      func(a []any) (any, error) { return UCF(a[0]), nil },
	"escre":    func(a []any) (any, error) { return EscRE(specStr(a[0])), nil },

	// a[1] is passed through as-is: Indent handles a float64 count itself,
	// so converting here would hide whether it does.
	"indent": func(a []any) (any, error) {
		return Indent(specStr(a[0]), a[1]), nil
	},

	"isbinext": func(a []any) (any, error) { return IsBinExt(specStr(a[0])), nil },
	"isbincontent": func(a []any) (any, error) {
		return IsBinContent([]byte(specStr(a[0]))), nil
	},

	"get":  func(a []any) (any, error) { return Get(a[0], specStr(a[1])), nil },
	"getx": func(a []any) (any, error) { return GetX(a[0], a[1]), nil },

	"deep": func(a []any) (any, error) { return Deep(a[0], a[1:]...), nil },

	"omap": func(a []any) (any, error) {
		m, _ := a[0].(map[string]any)
		return OMap(m), nil
	},

	"template": func(a []any) (any, error) {
		var spec *TemplateSpec
		if 3 <= len(a) && a[2] != nil {
			raw, ok := a[2].(map[string]any)
			if !ok {
				return nil, fmt.Errorf("template spec is not an object: %v", a[2])
			}
			spec = &TemplateSpec{}
			if rep, ok := raw["replace"].(map[string]any); ok {
				spec.Replace = rep
			}
		}
		return Template(specStr(a[0]), a[1], spec)
	},

	"names": func(a []any) (any, error) {
		base, _ := a[0].(map[string]any)
		if base == nil {
			base = map[string]any{}
		}
		if 3 == len(a) {
			return Names(base, specStr(a[1]), specStr(a[2])), nil
		}
		return Names(base, specStr(a[1])), nil
	},

	"lines": func(a []any) (any, error) { return Lines(specStr(a[0])), nil },
	"lcs": func(a []any) (any, error) {
		return LCS(specStrs(a[0]), specStrs(a[1])), nil
	},
}

// specStr coerces a corpus cell to string. A JSON null reaches a string
// parameter only where the corpus means the empty string.
func specStr(v any) string {
	if v == nil {
		return ""
	}
	if s, ok := v.(string); ok {
		return s
	}
	return fmt.Sprint(v)
}

func specStrs(v any) []string {
	items, ok := v.([]any)
	if !ok {
		return nil
	}
	out := make([]string, 0, len(items))
	for _, it := range items {
		out = append(out, specStr(it))
	}
	return out
}

// specCanon renders a value as canonical JSON: object keys sorted (which
// json.Marshal already does for maps), HTML escaping off so <, > and &
// survive as themselves the way JSON.stringify leaves them.
func specCanon(v any) (string, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(specNorm(v)); err != nil {
		return "", err
	}
	return strings.TrimRight(buf.String(), "\n"), nil
}

// specNorm flattens Go's typed containers to the plain shapes the corpus
// speaks, so a []string, a [][2]any and a []any all serialize alike.
//
// A nil slice or map becomes empty rather than null. Go draws a
// distinction there that the corpus does not: len 0 either way, and no
// caller can tell them apart without reflection. A genuinely absent
// value is an untyped nil and still serializes as null.
func specNorm(v any) any {
	if v == nil {
		return nil
	}

	rv := reflect.ValueOf(v)
	switch rv.Kind() {

	case reflect.Slice, reflect.Array:
		if reflect.Slice == rv.Kind() && rv.IsNil() {
			return []any{}
		}
		out := make([]any, rv.Len())
		for i := 0; i < rv.Len(); i++ {
			out[i] = specNorm(rv.Index(i).Interface())
		}
		return out

	case reflect.Map:
		if rv.IsNil() {
			return map[string]any{}
		}
		out := make(map[string]any, rv.Len())
		for _, k := range rv.MapKeys() {
			out[specStr(k.Interface())] = specNorm(rv.MapIndex(k).Interface())
		}
		return out

	case reflect.Ptr:
		if rv.IsNil() {
			return nil
		}
		return specNorm(rv.Elem().Interface())
	}

	return v
}

func loadSpecCases(t *testing.T) []specCase {
	t.Helper()

	entries, err := os.ReadDir(specDir)
	if err != nil {
		t.Fatalf("cannot read %s: %v", specDir, err)
	}

	names := []string{}
	for _, e := range entries {
		if strings.HasSuffix(e.Name(), ".tsv") {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)

	if 0 == len(names) {
		t.Fatalf("no .tsv files in %s", specDir)
	}

	out := []specCase{}
	wantHeader := []string{"id", "fn", "args", "expect", "error"}

	for _, name := range names {
		raw, err := os.ReadFile(filepath.Join(specDir, name))
		if err != nil {
			t.Fatalf("cannot read %s: %v", name, err)
		}

		// Corpus files are committed with LF; tolerate CRLF in case a
		// checkout or an editor rewrote them.
		text := strings.ReplaceAll(string(raw), "\r\n", "\n")
		header := false

		for i, row := range strings.Split(text, "\n") {
			if "" == strings.TrimSpace(row) || strings.HasPrefix(row, "#") {
				continue
			}

			cells := strings.Split(row, "\t")

			if !header {
				if !reflect.DeepEqual(cells, wantHeader) {
					t.Fatalf("%s: header is %v, want %v", name, cells, wantHeader)
				}
				header = true
				continue
			}

			// A trailing empty cell may be trimmed by an editor; pad
			// rather than index out of range.
			for len(cells) < 5 {
				cells = append(cells, "")
			}
			if 5 != len(cells) {
				t.Fatalf("%s:%d: %d cells, want 5", name, i+1, len(cells))
			}

			c := specCase{
				file:   name,
				line:   i + 1,
				id:     cells[0],
				fn:     cells[1],
				errSub: cells[4],
			}

			if err := json.Unmarshal([]byte(cells[2]), &c.args); err != nil {
				t.Fatalf("%s:%d %s: bad args JSON: %v", name, i+1, c.id, err)
			}
			if "" == c.errSub {
				if err := json.Unmarshal([]byte(cells[3]), &c.expect); err != nil {
					t.Fatalf("%s:%d %s: bad expect JSON: %v", name, i+1, c.id, err)
				}
			}

			out = append(out, c)
		}
	}

	return out
}

// A corpus that shrank to nothing would pass every assertion, so the
// count is checked too. The TS runner prints its own count; the two
// should agree.
func TestSpecCorpusLoaded(t *testing.T) {
	cases := loadSpecCases(t)
	if 100 >= len(cases) {
		t.Fatalf("only %d cases loaded from %s", len(cases), specDir)
	}
	t.Logf("spec corpus: %d cases", len(cases))
}

// An unknown fn fails rather than skipping. A corpus entry that one stack
// quietly ignores is precisely the divergence this suite exists to catch.
func TestSpecAllFnsDispatched(t *testing.T) {
	seen := map[string]bool{}
	for _, c := range loadSpecCases(t) {
		if _, ok := specFns[c.fn]; !ok {
			seen[c.fn] = true
		}
	}
	if 0 < len(seen) {
		missing := []string{}
		for fn := range seen {
			missing = append(missing, fn)
		}
		sort.Strings(missing)
		t.Fatalf("corpus uses undispatched fns: %s", strings.Join(missing, ", "))
	}
}

func TestSpecIDsUnique(t *testing.T) {
	seen := map[string]bool{}
	for _, c := range loadSpecCases(t) {
		key := c.file + ":" + c.id
		if seen[key] {
			t.Errorf("duplicate id %s", key)
		}
		seen[key] = true
	}
}

func TestSpecCorpus(t *testing.T) {
	for _, c := range loadSpecCases(t) {
		c := c
		t.Run(c.file+"/"+c.id, func(t *testing.T) {
			where := fmt.Sprintf("%s:%d %s", c.file, c.line, c.id)

			fn, ok := specFns[c.fn]
			if !ok {
				t.Fatalf("%s: fn %q is not dispatched", where, c.fn)
			}

			got, err := fn(c.args)

			if "" != c.errSub {
				if err == nil {
					t.Fatalf("%s: expected an error containing %q, got %v",
						where, c.errSub, got)
				}
				if !strings.Contains(err.Error(), c.errSub) {
					t.Fatalf("%s: error %q does not contain %q",
						where, err.Error(), c.errSub)
				}
				return
			}

			if err != nil {
				t.Fatalf("%s: unexpected error: %v", where, err)
			}

			gotJSON, gerr := specCanon(got)
			if gerr != nil {
				t.Fatalf("%s: cannot serialize result: %v", where, gerr)
			}
			wantJSON, werr := specCanon(c.expect)
			if werr != nil {
				t.Fatalf("%s: cannot serialize expectation: %v", where, werr)
			}

			if gotJSON != wantJSON {
				t.Errorf("%s\n  got  %s\n  want %s", where, gotJSON, wantJSON)
			}
		})
	}
}
